File: Airwallex/Core/Sources/Internal/AWXCountry.m
//
//  AWXCountry.m
//  Airwallex
//
//  Created by Victor Zhu on 2020/1/19.
//  Copyright © 2020 Airwallex. All rights reserved.
//

#import "AWXCountry.h"

@implementation AWXCountry

@end

@implementation AWXCountry (Utils)

+ (NSArray <AWXCountry *> *)allCountries
{
    NSLocale *locale = [NSLocale currentLocale];
    NSArray *isoCountryCodes = [NSLocale ISOCountryCodes];
    NSMutableArray *countries = [[NSMutableArray alloc] init];
    for (NSString *countryCode in isoCountryCodes) {
        NSString *countryName = [locale displayNameForKey:NSLocaleCountryCode value:countryCode];
        AWXCountry *country = [AWXCountry new];
        country.countryCode = countryCode;
        country.countryName = countryName;
        [countries addObject:country];
    }
    [countries sortUsingComparator:^NSComparisonResult(AWXCountry * _Nonnull obj1, AWXCountry * _Nonnull obj2) {
        return [obj1.countryName localizedCompare:obj2.countryName];
    }];
    return countries;
}

+ (nullable AWXCountry *)countryWithCode:(NSString *)code
{
    NSArray *filtered = [[self allCountries] filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"countryCode == %@", code]];
    return filtered.firstObject;
}

@end

File: Airwallex/Core/Sources/Internal/Models/AWXCountry.swift
//
//  AWXCountry.swift
//  Core
//
//  Created by Tony He (CTR) on 2024/7/26.
//  Copyright © 2024 Airwallex. All rights reserved.
//

import Foundation

/// `AWXCountry` includes the information of a country.
@objcMembers
@objc
public class AWXCountry: NSObject, Codable {
    /**
     Country code.
     */
    public let countryCode: String

    /**
     Country name.
     */
    public let countryName: String

    public init(countryCode: String, countryName: String) {
        self.countryCode = countryCode
        self.countryName = countryName
    }

    /**
     Return all of the supported countries.
     */
    public static func allCountries() -> [AWXCountry] {
        let locale = Locale.current
        let isoCountryCodes = Locale.isoRegionCodes
        var countries = [AWXCountry]()
        for isoCountryCode in isoCountryCodes {
            if let name = locale.localizedString(forCurrencyCode: isoCountryCode) {
                let country = AWXCountry(countryCode: isoCountryCode, countryName: name)
                countries.append(country)
            }
        }
        countries.sort {
            $0.countryName.localizedCompare($1.countryName) == .orderedAscending
        }

        return countries
    }

    /**
     Get a matched country object.

     @param code Country code.
     @return A country object.
     */
    public static func countryWithCode(_ code: String) -> AWXCountry? {
        return allCountries().first {
            $0.countryCode == code
        }
    }
}
